drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,295 @@
+ */
+static int32_t iio_ad7616_debug_reg_write(void *dev, uint32_t reg,
+		uint32_t val)
+{
+	return ad7616_write(dev, (uint8_t)reg, (uint16_t)val);
+}
+
+void iio_ad7616_event_handler(struct iio_desc *desc)
+{
+	iio_step(desc);
+}
+
+static int get_raw(void *device,
+		   char *buf,
+		   uint32_t len,
+		   const struct iio_ch_info *channel,
+		   intptr_t id)
+{
+	int ret;
+	uint8_t index = channel->ch_num < 8 ? channel->ch_num : channel->ch_num - 8;
+	struct ad7616_conversion_result res[8];
+	struct ad7616_dev *dev = device;
+
+	// TODO: we call this on every call, fetching all 16 inputs. Doesn't look smart
+	ret = ad7616_read_data_serial(dev, res, 1);
+	if (ret)
+		return ret;
+
+	if (channel->ch_num < 8)
+		return sprintf(buf, "%i", res[index].channel_a);
+	else
+		return sprintf(buf, "%i", res[index].channel_b);
+}
+
+static int set_raw(void *device,
+		   char *buf,
+		   uint32_t len,
+		   const struct iio_ch_info *channel,
+		   intptr_t id)
+{
+	/* NA- Can't set raw value */
+	return -EACCES;
+}
+
+static int get_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	struct ad7616_dev *dev = device;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	enum ad7616_range range = (ch < AD7616_VB0) ? dev->va[ch] : dev->vb[ch -
+				  AD7616_VB0];
+
+	switch (range) {
+	case AD7616_10V:
+		return sprintf(buf, "10");
+	case AD7616_5V:
+		return sprintf(buf, "5");
+	case AD7616_2V5:
+		return sprintf(buf, "2.5");
+	default:
+		return -EINVAL;
+	}
+}
+
+static int set_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_dev *dev = device;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	uint16_t ret;
+	struct ad7616_conversion_result results[8];
+	uint32_t bytes_per_scan = iio_dev_data->buffer->bytes_per_scan;
+	uint16_t data[2];
+	uint32_t i = 0;
+
+	for (; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(iio_dev_data->dev, results, 1);
+		if (ret)
+			return ret;
+
+		data[0] = results[0].channel_a;
+		data[1] = results[0].channel_b;
+
+		ret = no_os_cb_write(iio_dev_data->buffer->buf, data, bytes_per_scan);
```

Comment:
you should not call no_os_cb related APIs, instead please use iio_buffer_push_scan for each sample